File: nodejs/test/e2e/external-tool-cancellation.e2e.test.ts
/*---------------------------------------------------------------------------------------------
 *  Copyright (c) Microsoft Corporation. All rights reserved.
 *--------------------------------------------------------------------------------------------*/

import { describe, expect, it } from "vitest";
import { z } from "zod";
import { approveAll, defineTool } from "../../src/index.js";
import { createSdkTestContext } from "./harness/sdkTestContext.js";

describe("External tool cancellation", async () => {
    const { copilotClient: client } = await createSdkTestContext();

    async function withTimeout<T>(promise: Promise<T>, ms: number, label: string): Promise<T> {
        let timer: ReturnType<typeof setTimeout> | undefined;
        try {
            return await Promise.race([
                promise,
                new Promise<T>((_, reject) => {
                    timer = setTimeout(() => reject(new Error(`Timeout: ${label}`)), ms);
                }),
            ]);
        } finally {
            if (timer) clearTimeout(timer);
        }
    }

    it("should cancel tool handler when session disconnects", { timeout: 120_000 }, async () => {
        let toolStartedResolve!: () => void;
        const toolStarted = new Promise<void>((resolve) => {
            toolStartedResolve = resolve;
        });
        let toolCancelledResolve!: () => void;
        const toolCancelled = new Promise<void>((resolve) => {
            toolCancelledResolve = resolve;
        });
        let releaseToolResolve!: () => void;
        const releaseTool = new Promise<void>((resolve) => {
            releaseToolResolve = resolve;
        });

        const session = await client.createSession({
            onPermissionRequest: approveAll,
            tools: [
                defineTool("slow_analysis", {
                    description: "A slow analysis tool that blocks until released",
                    parameters: z.object({
                        value: z.string().describe("Value to analyze"),
                    }),
                    handler: async (_args, invocation) => {
                        toolStartedResolve();
                        await Promise.race([
                            releaseTool,
                            new Promise<never>((_, reject) =>
                                setImmediate(() => {
                                    const onAbort = () => {
                                        toolCancelledResolve();
                                        reject(new Error("aborted"));
                                    };
                                    if (invocation.signal?.aborted) {
                                        onAbort();
                                        return;
                                    }
                                    invocation.signal?.addEventListener("abort", onAbort, {
                                        once: true,
                                    });
                                })
                            ),
                        ]);
                        return "RELEASED";
                    },
                }),
            ],
        });

        try {
            void session.send({
                prompt: "Use slow_analysis with value 'test_abort'. Wait for the result.",
            });

            await withTimeout(toolStarted, 60_000, "slow_analysis start");
            await session.disconnect();
            await withTimeout(toolCancelled, 60_000, "slow_analysis cancellation");
        } finally {
            releaseToolResolve();
        }
    });
});

File: go/internal/e2e/external_tool_cancellation_e2e_test.go
package e2e

import (
	"testing"
	"time"

	copilot "github.com/github/copilot-sdk/go"
	"github.com/github/copilot-sdk/go/internal/e2e/testharness"
)

func TestExternalToolCancellationE2E(t *testing.T) {
	ctx := testharness.NewTestContext(t)
	client := ctx.NewClient()
	t.Cleanup(func() { client.ForceStop() })

	t.Run("should_cancel_tool_handler_when_session_disconnects", func(t *testing.T) {
		ctx.ConfigureForTest(t)

		type ValueParams struct {
			Value string `json:"value" jsonschema:"Value to analyze"`
		}
		toolStarted := make(chan struct{}, 1)
		toolCancelled := make(chan struct{}, 1)
		releaseTool := make(chan string, 1)

		slowTool := copilot.DefineTool("slow_analysis", "A slow analysis tool that blocks until released",
			func(_ ValueParams, inv copilot.ToolInvocation) (string, error) {
				select {
				case toolStarted <- struct{}{}:
				default:
				}
				select {
				case value := <-releaseTool:
					return value, nil
				case <-inv.TraceContext.Done():
					select {
					case toolCancelled <- struct{}{}:
					default:
					}
					return "", inv.TraceContext.Err()
				}
			})
		slowTool.SkipPermission = true

		session, err := client.CreateSession(t.Context(), &copilot.SessionConfig{
			OnPermissionRequest: copilot.PermissionHandler.ApproveAll,
			Tools:               []copilot.Tool{slowTool},
		})
		if err != nil {
			t.Fatalf("Failed to create session: %v", err)
		}
		t.Cleanup(func() { _ = session.Disconnect() })

		go func() {
			_, _ = session.Send(t.Context(), copilot.MessageOptions{
				Prompt: "Use slow_analysis with value 'test_abort'. Wait for the result.",
			})
		}()

		select {
		case <-toolStarted:
		case <-time.After(60 * time.Second):
			t.Fatal("Timed out waiting for tool handler to start")
		}

		if err := session.Disconnect(); err != nil {
			t.Fatalf("Disconnect failed: %v", err)
		}

		select {
		case <-toolCancelled:
		case <-time.After(60 * time.Second):
			t.Fatal("Timed out waiting for tool handler cancellation")
		}

	})
}
